### DISCUSSION TITLE
zustand mock issue with react-native and jest

### DISCUSSION THREAD

**Discussion by teampazi** [pmndrs/zustand#1970] [Bug report] [answered]
Hello,
I am very new to Zustand and i am working on a project using react-native. However, I am not able to test my component that utilizes zustand as the store manager because zustand is not properly initializing my state.

I created zustand.ts as described in the documentation with the following content

```
// zustand.ts
import * as zustand from "zustand"
  import { act } from "@testing-library/react-native";

  const { create: actualCreate } = jest.requireActual<typeof zustand.create>("zustand");

  export const storeResetFns = new Set<() => void>();

  export const create =  <T>(createState: zustand.StateCreator<T> | undefined) => () => {
      return typeof createState === "function" ? createInternalFn<T>(createState)  : createInternalFn;
  }

  export const createInternalFn = <S>(createState: zustand.StateCreator<S> | undefined): S => {
    const store =  typeof createState === "function"  ? actualCreate(createState) : actualCreate;
    const initialState = store.getState();
    storeResetFns.add(() => store.setState(initialState, true));
    return store;
  };

afterEach(() => {
    act(() => storeResetFns.forEach((resetFn) => resetFn()));
  });
```

And here is my actual state

```
// store.ts
import { create, GetState, StoreApi } from 'zustand';
  import { persist, createJSONStorage } from 'zustand/middleware';
  import AsyncStorage from '@react-native-async-storage/async-storage';

  import { StoreType } from 'path-to-type';
  import { storeSlice } from 'path-to-slice';

  export type AppStore = StoreType;
  const useStore = create<AppStore>(
    persist<StoreType>(
      (set, get, api) => ({
        ...storeSlice(
          set,
          get as GetState<StoreType>,
          api as StoreApi<StoreType>
        ),
      }),
      {
        name: 'storageKey',
        storage: createJSONStorage(() => AsyncStorage),
      }
    )
  );
  export { useStore };
```

But every time i run the test this is now what the store would look like

```
[Function: useBoundStore] {
      setState: [Function (anonymous)],
      getState: [Function: getState],
      subscribe: [Function: subscribe],
      destroy: [Function: destroy],
      persist: {
        setOptions: [Function: setOptions],
        clearStorage: [Function: clearStorage],
        getOptions: [Function: getOptions],
        rehydrate: [Function: rehydrate],
        hasHydrated: [Function: hasHydrated],
        onHydrate: [Function: onHydrate],
        onFinishHydration: [Function: onFinishHydration]
      }
    }
```

 while the store is expected to look  like 

```
{
      auth: {},
      setAuth: [Function: setAuthUser],
      deleteAuth: [Function: deleteAuth]
}
```

I could call `getState` and get the state in the test file and try to set the state but it doesn't reflect in the store

Please what am i doing wrong here please somebody help

**Comment by dbritto-dev** [comment]
@teampazi did you follow our testing guide?

**Reply by teampazi** [reply]
@dbritto-dev yes i did

**Reply by dbritto-dev** [reply]
@teampazi but you need to use the curried version on ts

**Reply by teampazi** [reply]
@dbritto-dev I did it does make any difference it basically means i have to rewrite the create to look like this
```
export const create =  () => <T>(createState: zustand.StateCreator<T> | undefined) => () => {
      return typeof createState === "function" ? createInternalFn<T>(createState)  : createInternalFn;
  }
```
otherwise it throws this error `TypeError: store.getState is not a function`
In the end it returns the same result.

**Reply by dbritto-dev** [reply]
@teampazi two things:

- You'd better using curried version on ts
- If you get `useBoundStore` as function is due to `Jest` not mocking `zustand` correctly or wrong mocking logic

In order to mock `zustand` you need to create `zustand.ts` file under `__mocks__` folder so `Jest` can pick it up

You can add some `console.log` statements in your mocking logic to validate if you are mocking `zustand` correctly

**Reply by teampazi** [reply]
@dbritto-dev this where I am confused i `mocked` it as described in the documentation inside the `__mocks__` folder also using the `curried` version and if i call `getState` in the `test file` or in the `mock` it returns the right value but i can't call getState in the actual `component` file because that's not how it's used. I believe jest is picking it up which is why the console log and `getState` returns the value inside the `zustand.ts` I believe it's something to do with the mocking logic, version of jest or the preset i am using which is `jest-expo` because it's an expo App. Does zustand work with `jest-expo`

**Reply by teampazi** [reply]
@dai-shi any ideas?

**Reply by dbritto-dev** [reply]
@teampazi would you mind creating a little demo on codesandbox? I would like to discard if there's something with React Native or with your mocking logic

**Reply by dai-shi** [reply]
so, does everything work fine for you without react-native? otherwise, if you try the same thing with codesandbox, @dbritto-dev would be able to review your logic.

**Reply by teampazi** [reply]
Started a sandbox here to mimic as close as possible what the structure of my code looks like https://codesandbox.io/p/sandbox/affectionate-grass-x2kmqh?file=%2Fjest.config.ts%3A13%2C29
However i haven't been able to get jest to run successfully but should be able to give you an idea what I am doing and if anything is wrong in the `zustand.ts` or the way the store is configured

**Reply by teampazi** [reply]
@dai-shi if you are asking if the code runs well on device yes it works fine only the test is not being mocked properly for reasons i am yet to find.

**Reply by teampazi** [reply]
I finally figured out what the Problem was. I was  doing 
```
export const create =  () => <T>(createState: zustand.StateCreator<T> | undefined) => () => {
      return typeof createState === "function" ? createInternalFn<T>(createState)  : createInternalFn;
  }
```
when i was using the curried version. However, It was supposed to be this
```
export const create =  () => <T>(createState: zustand.StateCreator<T> | undefined) => {
      return typeof createState === "function" ? createInternalFn<T>(createState)  : createInternalFn;
  }
```
with the curried version

**Reply by dbritto-dev** [reply]
@teampazi yeah that's why I asked you if you are following our testing guide, since you implemented your own version and that wasn't working well due to supporting curried and uncurried versions I encourage you to use always the curried version and use our code example in our testing guide

**Reply by dbritto-dev** [reply, marked as answer]
@teampazi yeah that's why I asked you if you are following our testing guide, since you implemented your own version and that wasn't working well due to supporting curried and uncurried versions I encourage you to use always the curried version and use the code example in our testing guide

**Reply by teampazi** [reply]
Thanks.
